File: leetcode_python/medium/BestTimeToBuyAndSellStockII.py
'''
Created on Dec 31, 2014

@author: Chandra Kant
'''
class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit(self, prices):
        if (prices == None or len(prices) == 0): return 0;
        ans = 0;
        running_sum = 0;
                
        for i in range(1, len(prices)) :
            diff = prices[i] - prices[i - 1];
            running_sum += diff;
            
            if running_sum < 0:
                running_sum = 0;
            if diff > 0:
                ans += diff;               
        return ans;
if __name__ == '__main__':
    sol = Solution();
    prices = [6, 7, 5, 7, 7, 10, 4, 8, 9, 14, 11, 2, 13];
    print(sol.maxProfit(prices));

File: leetcode/src/medium/BestTimeToBuyAndSellStockII.java
package medium;

public class BestTimeToBuyAndSellStockII {
	
	public int maxProfit(int[] prices) {
        if (prices == null || prices.length == 0) return 0;
        int maxProfit = 0;
        int runningSum = 0;
        
        for (int i = 1; i < prices.length; i++) {
        	int diff = prices[i] - prices[i - 1];
        	runningSum += diff;
        	if (runningSum < 0) {
        		runningSum = 0;
        	}
        	if (diff >= 0) {
        		maxProfit += diff;
        	}
        }        
        return maxProfit;
    }
	
	public static void main(String[] args) {
		BestTimeToBuyAndSellStockII sol = new BestTimeToBuyAndSellStockII();
		int[] prices = new int[] {6, 7, 5, 7, 7, 10, 4, 8, 9, 14, 11, 2, 13};
		System.out.println(sol.maxProfit(prices));
	}
}
